inference/enterprise/parallel/dispatch_manager.py

```diff
@@ -0,0 +1,173 @@
+import asyncio
+import json
+from dataclasses import dataclass
+from time import perf_counter, time
+from typing import Any, Coroutine, Optional
+
+from redis import Redis
+
+from inference.core.entities.requests.inference import (
+    InferenceRequest,
+    request_from_type,
+)
+from inference.core.entities.responses.inference import response_from_type
+from inference.core.env import NUM_PARALLEL_TASKS, REDIS_HOST, REDIS_PORT
+from inference.core.managers.base import ModelManager
+from inference.core.registries.roboflow import get_model_type
+from inference.enterprise.parallel.tasks import preprocess
+from inference.enterprise.parallel.utils import (
+    FAILURE_STATE,
+    INITIAL_STATE,
+    SUCCESS_STATE,
+    TASK_RESULT_KEY,
+    TASK_STATUS_KEY,
+)
+
+NOT_FINISHED_RESPONSE = "===NOTFINISHED==="
+
+
+class ResultsChecker:
+    """
+    Class responsible for queuing asyncronous inference runs,
+    keeping track of running requests, and awaiting their results.
+    """
+
+    def __init__(self):
+        self.tasks = []
+        self.dones = dict()
+        self.errors = dict()
+        self.running = True
+
+    def add_redis(self, r: Redis):
+        """
+        After instantiation, give results checker a redis object
+        """
+        self.r = r
+
+    async def add_task(self, t, request):
+        """
+        Wait until there's available cylce to queue a task.
+        When there are cycles, add the task's id to a list to keep track of its results,
+        launch the preprocess celeryt task, set the task's status to in progress in redis.
+        """
+        interval = 0.1
+        while len(self.tasks) > NUM_PARALLEL_TASKS:
+            await asyncio.sleep(interval)
+        self.tasks.append(t)
+        preprocess.s(request.dict()).delay()
+        self.r.set(TASK_STATUS_KEY.format(t), INITIAL_STATE)
+
+    def check_task(self, t: str) -> Any:
+        """
+        Check the done tasks and errored tasks for this task id.
+        """
+        if t in self.dones:
+            return self.dones.pop(t)
+        if t in self.errors:
+            message = self.errors.pop(t)
+            raise Exception(message)
+        return NOT_FINISHED_RESPONSE
+
+    async def loop(self):
+        """
+        Main loop. Check all in progress tasks for their status, and if their status is final,
+        (either failure or success) then add their results to the appropriate results dictionary.
+        """
+        interval = 0.1
+        while self.running:
+            tasks = [t for t in self.tasks]
```

Comment:
Does this code have a risk that the length of the array `self.tasks` changes while this list comprehension iterates over it? Wondering if we need to first make a copy of self.tasks with the `copy` module? Not totally sure though and if you haven't seen errors here before probably not a real problem.

Reply:
no, line 78 makes a copy!